en/news/_posts/2019-12-06-separation-of-positional-and-keyword-arguments-in-ruby-3-0.md

```diff
@@ -0,0 +1,377 @@
+
+You need to explicitly delegate keyword arguments.
+
+```
+def foo(*args, **kwargs, &block)
+  target(*args, **kwargs, &block)
+end
+```
+
+Alternatively, if you do not need compatibility with Ruby 2.6 or prior and you don't alter any arguments, you can use a delegation syntax (`...`) that is introduced in Ruby 2.7.
+
+```
+def foo(...)
+  target(...)
+end
+```
+
+### Ruby 2.7
+
+In short: use `Module#ruby2_keywords` and delegate `*args, &block`.
+
+```
+ruby2_keywords def foo(*args, &block)
+  target(*args, &block)
+end
+```
+
+`ruby2_keywords` accepts keyword arguments as the last Hash argument, and passes it as keyword arguments when calling the other method.
+
+In fact, Ruby 2.7 allows the new style of delegation in many cases.  However, there is a known corner case.  See the next section.
+
+### A compatible delegation that works on Ruby 2.6, 2.7 and Ruby 3
+
+In short: use `Module#ruby2_keywords` again.
+
+```
+ruby2_keywords def foo(*args, &block)
+  target(*args, &block)
+end
+```
+
+Unfortunately, we need to use the old-style delegation (i.e., no `**kwargs`) because Ruby 2.6 or prior does not handle the new delegation style correctly.  This is one of the reasons of the keyword argument separation; the details are described in the last section. And `ruby2_keywords` allows you to run the old style even in Ruby 2.7 and 3.0.  As there is no `ruby2_keywords` defined in 2.6 or prior, please use [ruby2_keywords gem](https://rubygems.org/gems/ruby2_keywords) or define it by yourself:
+
+```
+def ruby2_keywords(*)
+end if RUBY_VESRION < "2.7"
+```
+
+---
+
+If your code doesn't have to run on Ruby 2.6 or older, and you don't care a corner case, you may try the new style in Ruby 2.7.  In almost all cases, it works.  Note that, however, there are unfortunate corner cases as follows:
+
+```
+def target(*args)
+  p args
+end
+
+def foo(*args, **kwargs, &block)
+  target(*args, **kwargs, &block)
+end
+
+foo({})       #=> Ruby 2.7: []   ({} is dropped)
+foo({}, **{}) #=> Ruby 2.7: [{}] (You can pass {} by explicitly passing "no" keywords)
+```
+
+An empty Hash argument is automatically converted and absorbed to `**kwargs`, and the delegation call removes the empty keyword hash, so no argument is passed to `target`.  As far as we know, this is the only one corner case.
+
+As noted in the last line, you can work around this issue by using `**{}`.
+
+If you really worry about the portability, use `ruby2_keywords`.  (Acknowledge that Ruby 2.6 or before themselves have tons of corner cases in keyword arguments. :-)
+
+---
+
+In some extreme cases, you may need to check a Ruby version.  Typically, if accepting arguments is not done at the same time of passing them to another method, you need to write:
+
+```
+if RUBY_VERSION < "2.7"
+  def accept_and_save_args(*args, &block)
+    @args, @block = args, block
+  end
+
+  def pass_saved_args
+    target(*@args, &@block)
+  end
+else
+  ruby2_keywords def accept_and_save_args(*args, **kwargs, &block)
```

Comment:
This is incorrect.  You can't use `ruby2_keywords` on a method that accepts keyword arguments.

I would remove this example of checking Ruby version, since `ruby2_keywords def accept_and_save_args(*args, &block)` will correctly pass positional arguments and keywords to `target`.

I suppose you could remove `ruby2_keywords` from the else branch, but that results in significantly more code, and it is also slower in CRuby.

Reply:
Wow, I can't understand why I wrote this.  Removed at 213e70fb.

Do you have any example where a version check is needed?

Reply:
I suppose version checking could be needed if you are checking if the last argument is a hash, and using `Hash#merge`:

```ruby
def foo(*args)
  args[-1] = args[-1].merge(bar: false) if args[-1].is_a?(Hash)
  target(*args)
end
```

This wouldn't work with `ruby2_keywords` as the hash returned by `Hash#merge` does not have the flag set on it.  In this case you may want a version check, switching to using keywords on 2.7+.

Reply:
Thanks.  Well, I'm unsure what you want to do by this code.  It converts the calls:

* `foo(1, 2, 3)` to `bar(1, 2, 3)`
* `foo(1, 2, 3, {})` to `bar(1, 2, 3, {bar: false})`
* `foo(1, 2, 3, **{})` to `bar(1, 2, 3, bar: false)`

Is this really intended?

Reply:
It actually converts `foo(1, 2, 3, **{})` to `bar(1, 2, 3)`, since the double splat of empty hash does not pass an argument.

It would convert `foo(1, 2, 3, a: 'b')` to `bar(1, 2, 3, {a: 'b', bar: false})` (even with `ruby2_keywords`.  Under Ruby <2.7, the expectation is probably it would modify keywords and result in `bar(1, 2, 3, a: 'b', bar: false)`, but that doesn't happen as `Hash#merge` returns a hash without the keyword flag.  This is a possible case where you would want to use different method definitions, with explicit keywords on Ruby 2.7+ and the existing code on Ruby 2.6:

```ruby
if RUBY_VERSION < '2.7'
  def foo(*args)
    args[-1] = args[-1].merge(bar: false) if args[-1].is_a?(Hash)
    target(*args)
  end
else
  def foo(*args, **kw)
    target(*args, **kw, bar: false)
  end
end
```

Note that this does not result in the exact same behavior. In 2.6 this will modify a positional hash argument, while the 2.7 version does not.  Also, if `target` doesn't accept keywords, the 2.7+ code passes an additional `{bar: false}` argument to it.

This isn't a good example either, as the code given for 2.7 works as far back as 2.0.  So you should probably just switch to using keywords unless you need to support Ruby <2.0. 

So I can't think of a reasonable case where you would want to use a version check.  We should probably just not mention version checking.

Reply:
Thanks!

> So I can't think of a reasonable case where you would want to use a version check. We should probably just not mention version checking.

Agreed.  Let's add a reasonable case if/after we encounter it.